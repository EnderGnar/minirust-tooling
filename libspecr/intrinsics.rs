use std::ops::{Try, FromResidual, ControlFlow, Residual, Yeet};
use std::convert::Infallible;

#[derive(Clone, Hash, PartialEq, Eq, Debug)]
pub struct Name(pub String);
pub struct Nondet<T>(pub T);

pub macro yeet {
    ($x:expr) => {
        do yeet $x
    },
}

impl<T, E> Try for Nondet<Result<T, E>> {
    type Output = T;
    type Residual = Nondet<Result<Infallible, E>>;

    fn from_output(output: Self::Output) -> Self {
        Nondet(Ok(output))
    }

    fn branch(self) -> ControlFlow<Self::Residual, Self::Output> {
        match self.0 {
            Ok(x) => ControlFlow::Continue(x),
            Err(e) => ControlFlow::Break(Nondet(Err(e))),
        }
    }
}

impl<T, E> FromResidual<Nondet<Result<Infallible, E>>> for Nondet<Result<T, E>> {
    fn from_residual(residual: Nondet<Result<Infallible, E>>) -> Self {
        match residual.0 {
            Ok(x) => match x {},
            Err(e) => Nondet(Err(e))
        }
    }
}

impl<T, E> FromResidual<Result<Infallible, E>> for Nondet<Result<T, E>> {
    fn from_residual(residual: Result<Infallible, E>) -> Self {
        match residual {
            Ok(x) => match x {},
            Err(e) => Nondet(Err(e))
        }
    }
}

impl<T, E> Residual<T> for Nondet<Result<Infallible, E>> {
    type TryType = Nondet<Result<T, E>>;
}

impl<T, E> FromResidual<Yeet<E>> for Nondet<Result<T, E>> {
    fn from_residual(residual: Yeet<E>) -> Self {
        Nondet(Err(residual.0))
    }
}


impl<T> Try for Nondet<Option<T>> {
    type Output = T;
    type Residual = Nondet<Option<Infallible>>;

    fn from_output(output: Self::Output) -> Self {
        Nondet(Some(output))
    }

    fn branch(self) -> ControlFlow<Self::Residual, Self::Output> {
        match self.0 {
            Some(x) => ControlFlow::Continue(x),
            None => ControlFlow::Break(Nondet(None)),
        }
    }
}

impl<T> FromResidual<Nondet<Option<Infallible>>> for Nondet<Option<T>> {
    fn from_residual(residual: Nondet<Option<Infallible>>) -> Self {
        match residual.0 {
            Some(x) => match x {},
            None => Nondet(None)
        }
    }
}
