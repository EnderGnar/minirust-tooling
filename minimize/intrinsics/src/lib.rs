// These functions are is merely mocks.
// The actual implementation will be added by `minimize`.

pub fn print<T>(_t: T) { unreachable!() }
pub fn eprint<T>(_t: T) { unreachable!() }
pub fn exit() { unreachable!() }
