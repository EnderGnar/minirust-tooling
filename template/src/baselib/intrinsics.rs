use crate::baselib::*;

pub struct ArgAbi;
pub type BbName = String;
pub type LocalName = String;
pub type FnName = String;

pub struct Nondet<T>(pub(in crate::baselib) T);
