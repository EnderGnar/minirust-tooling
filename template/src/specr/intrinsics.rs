use crate::specr::*;

use std::ops::{Try, FromResidual, ControlFlow};
use std::convert::Infallible;

pub struct ArgAbi;
pub type BbName = String;
pub type LocalName = String;
pub type FnName = String;

pub struct Nondet<T>(pub(in crate::specr) T);

impl<T, E> Try for Nondet<Result<T, E>> {
    type Output = T;
    type Residual = Nondet<Result<Infallible, E>>;

    fn from_output(output: Self::Output) -> Self {
        Nondet(Ok(output))
    }

    fn branch(self) -> ControlFlow<Self::Residual, Self::Output> {
        match self.0 {
            Ok(x) => ControlFlow::Continue(x),
            Err(e) => ControlFlow::Break(Nondet(Err(e))),
        }
    }
}

impl<T, E> FromResidual<Nondet<Result<Infallible, E>>> for Nondet<Result<T, E>> {
    fn from_residual(residual: Nondet<Result<Infallible, E>>) -> Self {
        match residual.0 {
            Ok(x) => match x {},
            Err(e) => Nondet(Err(e))
        }
    }
}

impl<T, E> FromResidual<Result<Infallible, E>> for Nondet<Result<T, E>> {
    fn from_residual(residual: Result<Infallible, E>) -> Self {
        match residual {
            Ok(x) => match x {},
            Err(e) => Nondet(Err(e))
        }
    }
}
