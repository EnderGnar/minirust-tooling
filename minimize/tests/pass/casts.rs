extern crate intrinsics;
use intrinsics::*;

fn black_box<T>(t: T) -> T { t }

fn main() {
    print(black_box(0u32) as u8); // 0
    print(black_box(256u32 + 42) as u8); // 42
    print(black_box(256i16) as u8); // 0
    print(black_box(u8::MAX) as i8); // -1
    print(black_box(-1i8) as u8); // 255
    print(black_box(24u8) as i128); // 24

    let x = 2;
    let addr = &x as *const i32 as usize;
    print(addr - addr / 4 * 4); // 0, because addr%4 == 0 due to alignment.

    // TODO also test int to ptr casts when they are supported.
}
