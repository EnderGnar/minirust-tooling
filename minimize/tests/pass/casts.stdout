0
42
0
-1
255
24
0
